std/typecons.d

```diff
@@ -1787,6 +1798,155 @@ unittest
     assert(rebindable(pr3341_aa)[321] == 543);
 }
 
+/** Models safe reassignment of otherwise constant struct instances.
+ *
+ * A struct with a field of reference type cannot be assigned to a constant
+ * struct of the same type. `Rebindable!(const S)` allows assignment to
+ * a `const S` while enforcing only constant access to its fields.
+ *
+ * `Rebindable!(immutable S)` does the same but field access may create a
+ * temporary copy of `S` in order to enforce _true immutability.
+ */
+template Rebindable(S)
+if (is(S == struct))
+{
+    static if (isMutable!S)
+        alias Rebindable = S;
+    else
+    struct Rebindable
+    {
+        // fields of payload must be treated as tail const (unless S is mutable)
+        private Unqual!S payload;
+
+        this()(S s) @trusted
+        {
+            // we preserve tail immutable guarantees so cast is OK
+            payload = cast(Unqual!S)s;
```

Comment:
> Casting away const is UB in any situation.

Just casting is ok. You just can't mutate through the casted reference.

See https://dlang.org/spec/const3.html#removing_with_cast